logstash-core/src/main/java/org/logstash/plugin/Processor.java

```diff
@@ -0,0 +1,15 @@
+package org.logstash.plugin;
+
+import org.logstash.Event;
+
+import java.util.Collection;
+
+public interface Processor {
+    /**
+     * Process events. In the past, this was called a `filter` in Logstash.
+     *
+     * @param events The events to be processed
+     * @return Any new events created by this processor.
+     */
+    Collection<Event> process(Collection<Event> events);
```

Comment:
@jordansissel as discussed I think we should def not use `Collection` here.

What I really like would be this:

```java
public interface EventIterator {

    /**
     * @return Event currently pointed at
     */
    Event current();

    /**
     * @return True iff there was another Event in the iterator
     */
    boolean next();

    /**
     * @param buffer Collection to store all events in
     * @param count Maximum number of events to store in buffer
     * @return Number of events copied to buffer
     */
    int collect(Collection<Event> buffer, int count);
}
```

If we pass something like this down the callstack, we can hide the whole Queue logic behind it and each `Filter` and `Output` can decide if it wants to operate in a batched or non-batched way and we can optimize accordingly. Also we can simply signal end of execution by returning `false` from `next()` as well as `0` from `collect` (I think both can be made blocking, the current situation of returning empty batches under pressure is just wrong imo).

Reply:
I agree we can improve the object that an input uses to shuttle data into the pipeline. I think that is out of scope for this PR as it changes existing code.

After this PR (just the plugin interfaces  and any necessary support code) land, we can work on the next step of making Java plugins runnable in the pipeline -- and at that time, we can make the move to this EventIterator (or whatever is designed).

Reply:
@jordansissel yup makes sense, sadly enough the current Java execution is still `Collection` based.

The cool thing is that we can try out this approach internally in the Java execution first (make `RootDataset` and `OutputDataset` into adapters and then just remove both of them (those two are kind of awkward anyway, the `Dataset` approach only seems to be the correct abstraction for `If` and `Filter`) once it worked out).